### PR TITLE
SPARKC-111: Fixed bug in CassandraConnectorSource which caused problems when during initialization in executor environment

### PR DESCRIPTION
**Pull request by jacek-lewandowski**

### CODE REVIEW

File: spark-cassandra-connector/src/main/scala/org/apache/spark/metrics/CassandraConnectorSource.scala [hunk 1]
```diff
@@ -37,10 +38,7 @@ class CassandraConnectorSource extends Source {
 }
 
 object CassandraConnectorSource {
-  @volatile
-  private var _instance: Option[CassandraConnectorSource] = None
-  @volatile
-  private var _env: SparkEnv = null
+  private var _state: (Option[CassandraConnectorSource], SparkEnv) = (None, null)
```
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/metrics/CassandraConnectorSource.scala]: Why None and null in the same place?

File: spark-cassandra-connector/src/main/scala/org/apache/spark/metrics/CassandraConnectorSource.scala [hunk 2]
```diff
@@ -58,16 +58,41 @@ object CassandraConnectorSource {
 
   /** This method sets the given instance of [[CassandraConnectorSource]] as the current active instance.
     * It is allowed to call this method only once for the same `SparkEnv` instance which is currently
-    * up and running.
+    * up and running or this is the executor environment.
     */
   private def maybeSetInstance(ccs: CassandraConnectorSource): Unit = synchronized {
     val curEnv = SparkEnv.get
+    val (_instance, _env) = _state
     assert(ccs != null, "ccs cannot be null")
```
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/metrics/CassandraConnectorSource.scala]: Use require for precondition checking.

**Comment by pkolaczk** [comment]
Generally +1, however I don't like those singletons in Spark everywhere (SparkEnv and stuff). Smells bad design to me.

File: spark-cassandra-connector/src/main/scala/org/apache/spark/metrics/CassandraConnectorSource.scala (hunk 1 above)
Comment (pkolaczk) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/metrics/CassandraConnectorSource.scala]: You can use `Option.apply` to convert nulls / non-nulls to `None` / `Some` automatically.

File: spark-cassandra-connector/src/main/scala/org/apache/spark/metrics/CassandraConnectorSource.scala (hunk 2 above)
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/metrics/CassandraConnectorSource.scala]: I really meant assertion. This is not intended to be used internally only and valid use of it should never cause those cases to occur.

File: spark-cassandra-connector/src/main/scala/org/apache/spark/metrics/CassandraConnectorSource.scala (hunk 1 above)
Comment (jacek-lewandowski) [inline spark-cassandra-connector/src/main/scala/org/apache/spark/metrics/CassandraConnectorSource.scala]: ok

**Comment by jacek-lewandowski** [comment]
I don't like it either. However I have 0 ideas how to do this other way without modifying Spark.

**Pull request merged**